src/traces/indicator/plot.js

```diff
@@ -0,0 +1,788 @@
+            l: fullLayout._size.l + fullLayout._size.w * domain.x[0],
+            r: fullLayout._size.r + fullLayout._size.w * (1 - domain.x[1]),
+            t: fullLayout._size.t + fullLayout._size.h * (1 - domain.y[1]),
+            b: fullLayout._size.b + fullLayout._size.h * (domain.y[0])
+        };
+        var centerX = size.l + size.w / 2;
+        var centerY = size.t + size.h / 2;
+
+
+        // Angular gauge size
+        var radius = Math.min(size.w / 2, size.h); // fill domain
+        var innerRadius = cn.innerRadius * radius;
+
+        // Position numbers based on mode and set the scaling logic
+        var numbersX, numbersY, numbersScaler;
+        var numbersAlign = trace.align || 'center';
+
+        numbersY = centerY;
+        if(!hasGauge) {
+            numbersX = size.l + position[numbersAlign] * size.w;
+            numbersScaler = function(el) {
+                return fitTextInsideBox(el, 0.9 * size.w, 0.9 * size.h);
+            };
+        } else {
+            if(isAngular) {
+                numbersX = centerX;
+                numbersY = centerY + radius / 2;
+                numbersScaler = function(el) {
+                    return fitTextInsideCircle(el, 0.9 * innerRadius);
+                };
+            }
+            if(isBullet) {
+                var padding = cn.bulletPadding;
+                var p = (1 - cn.bulletNumberDomainSize) + padding;
+                numbersX = size.l + (p + (1 - p) * position[numbersAlign]) * size.w;
+                numbersScaler = function(el) {
+                    return fitTextInsideBox(el, (cn.bulletNumberDomainSize - padding) * size.w, size.h);
+                };
+            }
+        }
+
+        // Draw numbers
+        var numbersOpts = {
+            numbersX: numbersX,
+            numbersY: numbersY,
+            numbersScaler: numbersScaler,
+            hasTransition: hasTransition,
+            transitionOpts: transitionOpts,
+            onComplete: onComplete
+        };
+        drawNumbers(gd, plotGroup, cd, numbersOpts);
+
+        // Reexpress our gauge background attributes for drawing
+        var gaugeBg, gaugeOutline;
+        if(hasGauge) {
+            gaugeBg = {
+                range: [trace.vmin, trace.vmax],
+                color: trace.gauge.bgcolor,
+                line: {
+                    color: trace.gauge.bordercolor,
+                    width: 0
+                },
+                thickness: 1
+            };
+
+            gaugeOutline = {
+                range: [trace.vmin, trace.vmax],
+                color: 'rgba(0, 0, 0, 0)',
+                line: {
+                    color: trace.gauge.bordercolor,
+                    width: trace.gauge.borderwidth
+                },
+                thickness: 1
+            };
+        }
+
+        // Prepare angular gauge layers
+        var data = cd.filter(function() {return isAngular;});
+        var angularGauge = plotGroup.selectAll('g.angular').data(data);
```

Comment:
`cd` here has always `length===1`, correct?

In this case, 

```js
var angularGauge = plotGroup.selectAll('g.angular').data(isAngular ? cd : []);
```

would probably be more readable and make the (slightly confusing) `data` variable obsolete.

Reply:
Done in e6d8525ab07c56ca10bdbac6762b06d1d57bf830